pyriemann/datasets/sampling.py

```diff
@@ -42,6 +43,96 @@ def _pdf_r(r, sigma):
     return np.exp(partial_1 + partial_2)
 
 
+def _rejection_sampling_2D_gfunction_plus(sigma, r_sample):
+    """ auxiliary function used for the 2D rejection sampling
+    algorithm in the case where r is sampled with
+    the function g+.
+    Parameters
+    ----------
+    sigma : float
+        Dispersion of the Riemannian Gaussian distribution.
+    r_samples : ndarray, shape (1, n_dim)
+        Sample of the r parameters of the Riemannian Gaussian distribution.
+    Returns
+    -------
+    probability_of_acceptation : float
+    """
+    MU_A = np.array([sigma**2/2, -sigma**2/2])
+    COV_MATRIX = (sigma**2)*np.eye(2)
+    M = np.pi*(sigma**2)*np.exp(sigma**2/4)
+    if r_sample[0] >= r_sample[1]:
+        num = np.exp(-1/(2*sigma**2) * np.sum(r_sample**2))\
+            * np.sinh((r_sample[0] - r_sample[1])/2) * 1/M
+        den = multivariate_normal.pdf(r_sample, mean=MU_A, cov=COV_MATRIX)
+        return num / den
+    return 0
+
+
+def _rejection_sampling_2D_gfunction_minus(sigma, r_sample):
+    """ auxiliary function used for the 2D rejection sampling
+    algorithm in the case where r is sampled with
+    the function g-.
+    Parameters
+    ----------
+    sigma : float
+        Dispersion of the Riemannian Gaussian distribution.
+    r_samples : ndarray, shape (n_samples, n_dim)
+        Samples of the r parameters of the Riemannian Gaussian distribution.
+    Returns
+    -------
+    probability_of_acceptation : float
+    """
+    MU_B = np.array([-sigma**2/2, sigma**2/2])
+    COV_MATRIX = (sigma**2)*np.eye(2)
+    M = np.pi*(sigma**2)*np.exp(sigma**2/4)
+    if r_sample[0] < r_sample[1]:
+        num = np.exp(-1/(2*sigma**2) * np.sum(r_sample**2))\
+            * np.sinh((r_sample[1] - r_sample[0])/2)
+        den = multivariate_normal.pdf(r_sample, mean=MU_B, cov=COV_MATRIX)*M
+        return num/den
+    return 0
+
+
+def _rejection_sampling_2D(n_samples, sigma, random_state=None):
+    """ Rejection sampling algorithm for the 2D case. The algorithm is
+    optimized for spatial complexity but not for time complexity.
+    Works very well for low sigma values
+
+    Parameters
+    ----------
+    n_samples : int
+        Number of samples to get from the ptarget distribution.
+    sigma : float
+        Dispersion of the Riemannian Gaussian distribution.
+    random_state : int, RandomState instance or None, default=None
+        Pass an int for reproducible output across multiple function calls.
+    Returns
+    -------
+    r_samples : ndarray, shape (n_samples, n_dim)
+        Samples of the r parameters of the Riemannian Gaussian distribution.
+    """
+    MU_A = np.array([sigma**2/2, -sigma**2/2])
+    MU_B = np.array([-sigma**2/2, sigma**2/2])
+    COV_MATRIX = (sigma**2)*np.eye(2)
+    RES = []
+    cpt = 0
+    rs = check_random_state(random_state)
+    while cpt != n_samples:
```

Comment:
use a for loop not while here

Reply:
I would also prefer to use a for loop, but this would mean that we know the number of steps that we need to get `n_samples`... and this is not really the case, since we're rejecting candidates randomly. Of course, we could estimate a "reasonable" number of passes through the for loop like making it n_samples / prob_acceptance, but the prob_acceptance is hard to calculate and it is just an estimate of the number of passes we will go through anyways. Note that in the `sbi` package they do rejection_sampling with a while loop too; see https://github.com/mackelab/sbi/blob/main/sbi/samplers/rejection/rejection.py